cheetah/particles/parameter_beam.py

```
@@ -37,18 +37,14 @@ def __init__(
         factory_kwargs = {"device": device, "dtype": dtype}
         super().__init__()
 
-        self.register_buffer("_mu", None)
-        self.register_buffer("_cov", None)
-        self.register_buffer("energy", None)
-        self.register_buffer("total_charge", torch.tensor(0.0, **factory_kwargs))
+        self.species = species if species is not None else Species("electron")
 
-        self._mu = torch.as_tensor(mu, **factory_kwargs)
-        self._cov = torch.as_tensor(cov, **factory_kwargs)
-        self.energy = torch.as_tensor(energy, **factory_kwargs)
-        if total_charge is not None:
-            self.total_charge = torch.as_tensor(total_charge, **factory_kwargs)
+        total_charge = total_charge if total_charge is not None else torch.tensor(0.0)
 
-        self.species = species if species is not None else Species("electron")
+        self.register_buffer("_mu", mu.to(**factory_kwargs))
+        self.register_buffer("_cov", cov.to(**factory_kwargs))
+        self.register_buffer("energy", energy.to(**factory_kwargs))
+        self.register_buffer("total_charge", total_charge.to(**factory_kwargs))
```

The trouble with this style of initialization is that `nn.Parameter` are not properly registered since all properties are forced as buffers. That is, if you pass a `nn.Parameter` while creating a beam, torch will silently convert the parameter to a buffer. Creating a parametrized beam is still possible, but only _after_ initialization by assigning a `nn.Parameter` to a property.

---

So essentially, the previous way of initialization was a way to take advantage of property assignment magic that torch is performing in the background. But to be honest, I am not sure if the initial manual registration did anything useful or if we could just rely on the property assignment as long as we make sure to always assign tensors.

---

If we just assign a tensor as a property, it actually won't be registered as a buffer. I will run some speed tests with and without the change and then probably revert it. The property assignment magic seems to take some time ... but it is a feature that we want.

And either way, I'm still confused about whether the change actually made a difference or not. The measurements are contradictory 🤔